Make Duration safe for concurrent use

borgstrom opened the pull request:
The comment in `ForAttempt` says that it's safe for concurrent use, which it is, but I believe it is misleading because `Duration` is not safe for concurrent use.

If you checkout f452bc9 and run the tests with race detection it confirms this:

```
#550 ❯❯❯ go test -v -race -covermode atomic ./...
=== RUN   Test1
--- PASS: Test1 (0.00s)
=== RUN   TestForAttempt
--- PASS: TestForAttempt (0.00s)
=== RUN   Test2
--- PASS: Test2 (0.00s)
=== RUN   Test3
--- PASS: Test3 (0.00s)
=== RUN   Test4
--- PASS: Test4 (0.00s)
=== RUN   TestGetAttempt
--- PASS: TestGetAttempt (0.00s)
=== RUN   TestJitter
--- PASS: TestJitter (0.00s)
=== RUN   TestCopy
--- PASS: TestCopy (0.00s)
=== RUN   TestConcurrent
==================
WARNING: DATA RACE
Read at 0x00c000100060 by goroutine 16:
  _/Users/eborgstrom/Projects/backoff.(*Backoff).Duration()
      /Users/eborgstrom/Projects/backoff/backoff.go:28 +0x56
  _/Users/eborgstrom/Projects/backoff.TestConcurrent.func1()
      /Users/eborgstrom/Projects/backoff/backoff_test.go:134 +0x4a

Previous write at 0x00c000100060 by goroutine 15:
  _/Users/eborgstrom/Projects/backoff.(*Backoff).Duration()
      /Users/eborgstrom/Projects/backoff/backoff.go:29 +0xa6
  _/Users/eborgstrom/Projects/backoff.TestConcurrent.func1()
      /Users/eborgstrom/Projects/backoff/backoff_test.go:134 +0x4a

Goroutine 16 (running) created at:
  _/Users/eborgstrom/Projects/backoff.TestConcurrent()
      /Users/eborgstrom/Projects/backoff/backoff_test.go:140 +0x1d0
  testing.tRunner()
      /usr/local/opt/go/libexec/src/testing/testing.go:865 +0x163

Goroutine 15 (running) created at:
  _/Users/eborgstrom/Projects/backoff.TestConcurrent()
      /Users/eborgstrom/Projects/backoff/backoff_test.go:139 +0x1ba
  testing.tRunner()
      /usr/local/opt/go/libexec/src/testing/testing.go:865 +0x163
==================
--- FAIL: TestConcurrent (0.20s)
    testing.go:809: race detected during execution of test
FAIL
coverage: 79.3% of statements
FAIL    _/Users/eborgstrom/Projects/backoff     0.256s
```

The changes in 3e39e52 switch this package to use the [atomic](https://golang.org/pkg/sync/atomic/) package for loading, incrementing, and resetting the `attempts` variable such that it is now safe to use `Duration` from multiple goroutines.

jpillora commented:
Thanks for the PR

However, this alters the API so user programs would break.

Also I think the better way to make duration thread safe would be to use a mutex, as certain situations could return the same duration from concurrent calls

Lastly, I’m not convinced that duration needs to be thread safe. What’s your use case where you hit a race?

borgstrom commented:
> However, this alters the API so user programs would break.

Fair.  I have updated the change to avoid breaking the existing API signature.

> as certain situations could return the same duration from concurrent calls

Fixed.

> Lastly, I’m not convinced that duration needs to be thread safe. What’s your use case where you hit a race?

I didn't hit a race and don't have a use case.  This package was recently added to a project by a colleague and while reading the source code I found it dishonest that the documentation mentions the term "concurrent-safe" -- even if it's only mentioned in `ForAttempt` I can see someone misconstruing that to believe that calling `Duration` is "concurrent-safe" because its documentation says to "See ForAttempt".

Do what you will with it.

jpillora commented:
Thanks!
